From d219293d2cc560e7e939458c14d68ec4902d606a Mon Sep 17 00:00:00 2001
From: Peter Zijlstra <peterz@infradead.org>
Date: Tue, 24 Feb 2026 13:29:09 +0100
Subject: [PATCH 09/65] perf: Fix __perf_event_overflow() vs
 perf_remove_from_context() race

Make sure that __perf_event_overflow() runs with IRQs disabled for all
possible callchains. Specifically the software events can end up running
it with only preemption disabled.

This opens up a race vs perf_event_exit_event() and friends that will go
and free various things the overflow path expects to be present, like
the BPF program.

Fixes: 592903cdcbf6 ("perf_counter: add an event_list")
Reported-by: Simond Hu <cmdhh1767@gmail.com>
Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
Tested-by: Simond Hu <cmdhh1767@gmail.com>
Link: https://patch.msgid.link/20260224122909.GV1395416@noisy.programming.kicks-ass.net
---
 kernel/events/core.c | 62 +++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 56 insertions(+), 6 deletions(-)

diff --git a/kernel/events/core.c b/kernel/events/core.c
index ae482eb73ae15..83688e8d25822 100644
--- a/kernel/events/core.c
+++ b/kernel/events/core.c
@@ -9376,6 +9376,13 @@ int perf_event_overflow(struct perf_event *event,
 			  struct perf_sample_data *data,
 			  struct pt_regs *regs)
 {
+	/*
+	 * Entry point from hardware PMI, interrupts should be disabled here.
+	 * This serializes us against perf_event_remove_from_context() in
+	 * things like perf_event_release_kernel().
+	 */
+	lockdep_assert_irqs_disabled();
+
 	return __perf_event_overflow(event, 1, data, regs);
 }
 
@@ -9455,7 +9462,21 @@ static void perf_swevent_event(struct perf_event *event, u64 nr,
 			       struct pt_regs *regs)
 {
 	struct hw_perf_event *hwc = &event->hw;
+	unsigned long flags;
 
+	/*
+	 * This is:
+	 *   - software		preempt
+	 *   - tracepoint		preempt
+	 *   -   tp_target_task	irq (ctx->lock)
+	 *   - uprobes			preempt/irq
+	 *   - kprobes			preempt/irq
+	 *   - hw_breakpoint		irq
+	 *
+	 * Any of these are sufficient to hold off RCU and thus ensure @event
+	 * exists.
+	 */
+	lockdep_assert_preemption_disabled();
 	local64_add(nr, &event->count);
 
 	if (!regs)
@@ -9464,19 +9485,39 @@ static void perf_swevent_event(struct perf_event *event, u64 nr,
 	if (!is_sampling_event(event))
 		return;
 
+	/*
+	 * Serialize against event_function_call() IPIs like normal overflow
+	 * event handling. Specifically, must not allow
+	 * perf_event_release_kernel() -> perf_remove_from_context() to make
+	 * progress and 'release' the event from under us.
+	 */
+	local_irq_save(flags);
+	if (event->state != PERF_EVENT_STATE_ACTIVE) {
+		local_irq_restore(flags);
+		return;
+	}
+
 	if ((event->attr.sample_type & PERF_SAMPLE_PERIOD) && !event->attr.freq) {
 		data->period = nr;
-		return perf_swevent_overflow(event, 1, data, regs);
+		perf_swevent_overflow(event, 1, data, regs);
+		local_irq_restore(flags);
+		return;
 	} else
 		data->period = event->hw.last_period;
 
-	if (nr == 1 && hwc->sample_period == 1 && !event->attr.freq)
-		return perf_swevent_overflow(event, 1, data, regs);
+	if (nr == 1 && hwc->sample_period == 1 && !event->attr.freq) {
+		perf_swevent_overflow(event, 1, data, regs);
+		local_irq_restore(flags);
+		return;
+	}
 
-	if (local64_add_negative(nr, &hwc->period_left))
+	if (local64_add_negative(nr, &hwc->period_left)) {
+		local_irq_restore(flags);
 		return;
+	}
 
 	perf_swevent_overflow(event, 0, data, regs);
+	local_irq_restore(flags);
 }
 
 static int perf_exclude_event(struct perf_event *event,
@@ -9879,7 +9920,6 @@ void perf_tp_event(u16 event_type, u64 count, void *record, int entry_size,
 {
 	struct perf_sample_data data;
 	struct perf_event *event;
-
 	struct perf_raw_record raw = {
 		.frag = {
 			.size = entry_size,
@@ -9887,6 +9927,11 @@ void perf_tp_event(u16 event_type, u64 count, void *record, int entry_size,
 		},
 	};
 
+	/*
+	 * Per being a tracepoint, this runs with preemption disabled.
+	 */
+	lockdep_assert_preemption_disabled();
+
 	perf_sample_data_init(&data, 0, 0);
 	data.raw = &raw;
 
@@ -10328,6 +10373,11 @@ void perf_bp_event(struct perf_event *bp, void *data)
 	struct perf_sample_data sample;
 	struct pt_regs *regs = data;
 
+	/*
+	 * Exception context, will have interrupts disabled.
+	 */
+	lockdep_assert_irqs_disabled();
+
 	perf_sample_data_init(&sample, bp->attr.bp_addr, 0);
 
 	if (!bp->hw.state && !perf_exclude_event(bp, regs))
@@ -10781,7 +10831,7 @@ static enum hrtimer_restart perf_swevent_hrtimer(struct hrtimer *hrtimer)
 
 	if (regs && !perf_exclude_event(event, regs)) {
 		if (!(event->attr.exclude_idle && is_idle_task(current)))
-			if (__perf_event_overflow(event, 1, &data, regs))
+			if (perf_event_overflow(event, &data, regs))
 				ret = HRTIMER_NORESTART;
 	}
 
-- 
2.43.0

